### PR TITLE
Don't lint on executable crates with `non_snake_case` names

### LINKED ISSUE #45127: Don't enforce snake_case for binary names
Currently, if you have a Cargo.toml like this:

```toml
[package]
name = "some_package"
version = "0.0.0"

[[bin]]
name = "somePackage"
path = "src/main.rs"
```

rustc will complain:

```
warning: crate `somePackage` should have a snake case name such as `some_package`
  |
  = note: #[warn(non_snake_case)] on by default
```
I understand why this is reasonable for library crates, to enforce a common naming convention for rust libraries and source code. But it doesn't make much sense to enforce this for binaries as well, as there is no real advantage to doing so (note that the crate name isn't camelCase, only the binary name is).

---
```sh
$ rustc --version
rustc 1.22.0-nightly (05f8ddc46 2017-10-07)
```

### PR DESCRIPTION
**Pull request by jieyouxu**
Revives #111130, cc @GilShoshan94.
Closes #45127.

### CODE REVIEW

File: compiler/rustc_lint/src/nonstandard_style.rs [hunk 1]
```diff
@@ -366,7 +367,9 @@ impl<'tcx> LateLintPass<'tcx> for NonSnakeCase {
                 })
         };
 
-        if let Some(ident) = &crate_ident {
+        if let Some(ident) = &crate_ident
+            && cx.tcx.crate_types().iter().all(|&crate_type| crate_type != CrateType::Executable)
```
Comment (petrochenkov) [inline compiler/rustc_lint/src/nonstandard_style.rs:371]: This condition can be checked before evaluating `crate_ident`.

Also, it doesn't matter much, but I'd change the condition to `.any(|&crate_type| crate_type != CrateType::Executable)` to still warn on library+executable pairs.

Comment (jieyouxu) [reply compiler/rustc_lint/src/nonstandard_style.rs]: I hoisted the check before evaluating `crate_ident`, and made it so if the crate types are mixed library / executable we still check those.
@rustbot ready

**Comment by matthiaskrgr** [comment]
https://github.com/rust-lang/rust/pull/121886#issuecomment-1974690445
@bors r-

**Comment by jieyouxu** [comment]
Added `//@ ignore-wasm` to the dylib/proc-macro crate types.
@rustbot ready

**Comment by jieyouxu** [comment]
@petrochenkov I added `//@ ignore-i686-unknown-linux-musl` to the cdylib test. But this seems like whack-a-mole. Even after this ignore, I think it's likely the PR will fail in various other targets that don't support various crate-types. Do you think it's sufficient to test it for a single target (like some tier 1 linux/msvc) since we're just testing if this lint is not fired for executable-only cases?
@rustbot ready

**Pull request merged**